File: 2021/gentzen.rs
use std::fmt;
use std::clone::Clone;
use std::convert::identity;

#[derive(Clone)]
enum Proof<T: fmt::Debug> {
    Proof(Box<T>),
}

impl<T: fmt::Debug> fmt::Debug for Proof<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Proof::Proof(x) => write!(f, "⊢ {:?}", x),
        }
    }
}

fn from_proof<T: fmt::Debug>(x : Proof<T>) -> T {
    match x {
        Proof::Proof(x) => match *x { x => x },
    }
}

fn mk_proof<T: fmt::Debug>(x : T) -> Proof<T> {
    Proof::Proof(Box::new(x))
}

#[derive(Clone)]
enum PropCalc<T: fmt::Debug> {
    PropVar(Box<T>),
    Not(Box<PropCalc<T>>),
    And(Box<PropCalc<T>>, Box<PropCalc<T>>),
//    Or(Box<PropCalc<T>>, Box<PropCalc<T>>),
    Imp(Box<PropCalc<T>>, Box<PropCalc<T>>),
}

impl<T: fmt::Debug> fmt::Debug for PropCalc<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            PropCalc::PropVar(x) => write!(f, "{:?}", x),
            PropCalc::Not(x) => write!(f, "¬({:?})", x),
            PropCalc::And(x, y) => write!(f, "({:?}∧{:?})", x, y),
//            PropCalc::Or(x, y) => write!(f, "({:?}∨{:?})", x, y),
            PropCalc::Imp(x, y) => write!(f, "({:?}→{:?})", x, y),
        }
    }
}

fn tm_var<T: fmt::Debug>(x : T) -> PropCalc<T> {
    PropCalc::PropVar(Box::new(x))
}

fn tm_not<T: fmt::Debug>(x : PropCalc<T>) -> PropCalc<T> {
    PropCalc::Not(Box::new(x))
}

fn tm_and<T: fmt::Debug>(x : PropCalc<T>, y : PropCalc<T>) -> PropCalc<T> {
    PropCalc::And(Box::new(x), Box::new(y))
}

fn tm_imp<T: fmt::Debug>(x : PropCalc<T>, y : PropCalc<T>) -> PropCalc<T> {
    PropCalc::Imp(Box::new(x), Box::new(y))
}

fn rule_fantasy<T: fmt::Debug+Clone>(x : PropCalc<T>, y : &dyn Fn(Proof<PropCalc<T>>) -> Proof<PropCalc<T>>) -> Proof<PropCalc<T>> {
    let a = x;
    let b = a.clone();
    return mk_proof(tm_imp(a, from_proof(y(mk_proof(b)))));
}

fn rule_join<T: fmt::Debug>(x : Proof<PropCalc<T>>, y : Proof<PropCalc<T>>) -> Proof<PropCalc<T>> {
    mk_proof(tm_and(from_proof(x), from_proof(y)))
}

fn rule_double_tilde_intro<T: fmt::Debug>(x : Proof<PropCalc<T>>) -> Proof<PropCalc<T>> {
    mk_proof(tm_not(tm_not(from_proof(x))))
}

fn rule_double_tilde_elim<T: fmt::Debug>(x : Proof<PropCalc<T>>) -> Proof<PropCalc<T>> {
    match from_proof(x) {
        PropCalc::Not(boxed1) => match *boxed1 {
            PropCalc::Not(boxed2) => mk_proof(*boxed2),
            x => mk_proof(tm_not(x))
        },
        x => mk_proof(x),
    }
}

fn main() {
    let prop1_1 = tm_var('A');
    //let thm1 = rule_fantasy(prop1_1, &|i| -> Proof<PropCalc<char>> { i });
    let thm1 = rule_fantasy(prop1_1, &identity);
    println!("{:?}", thm1);

    let thm2 = rule_double_tilde_intro(thm1);
    println!("{:?}", thm2);

    let thm3 = rule_double_tilde_elim(thm2);
    println!("{:?}", thm3);

    let prop4_1 = tm_var('P');
    let prop4_2 = tm_var('Q');
    let prop4_3 = tm_and(prop4_1, prop4_2);
    let thm4 = rule_fantasy(prop4_3, &|i| -> Proof<PropCalc<char>> { let j = i.clone(); rule_join(i, j) });
    println!("{:?}", thm4);
}

File: 2021/gentzen.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// To check leaks, first compile:
// gcc gentzen.c && ./a.out
//
// And then:
// leaks -atExit -- ./a.out | grep LEAK:
//
// Debug info:
// gcc -g -fsanitize=address gentzen.c && ./a.out

typedef struct BinaryOp BinaryOp;
typedef struct UnaryOp UnaryOp;
typedef struct PropCalc PropCalc;
typedef struct Proof Proof;

// PropCalc and Proof structures
struct BinaryOp {
	PropCalc *l;
	PropCalc *r;
};

struct UnaryOp {
	PropCalc *v;
};

struct PropCalc {
	enum { Var, Not, And, Or, Imp } tag;
	union {
		char var;
		UnaryOp unop;
		BinaryOp binop;
	} data;
};

struct Proof {
	PropCalc *x;
};

// Proof helpers
Proof *mk_proof(PropCalc *a) {
	Proof *prf = (Proof *)malloc(sizeof(Proof));
	prf->x = a;
	return prf;
}

PropCalc *from_proof(Proof *a) {
	return a->x;
}

// Memory helpers
void free_propcalc(PropCalc *x) {
	if (x->tag == Not) {
		free_propcalc(x->data.unop.v);
	} else if (x->tag != Var) {
		free_propcalc(x->data.binop.l);
		free_propcalc(x->data.binop.r);
	}
	free(x);
}

void free_proof(Proof *x) {
	free_propcalc(x->x);
	free(x);
}

// Printers
void _print_propcalc(PropCalc *x) {
	switch (x->tag) {
		case Var:
			printf("%c", x->data.var);
			break;
		case Not:
			printf("!");
			_print_propcalc(x->data.unop.v);
			break;
		case And:
			_print_propcalc(x->data.binop.l);
			printf("&");
			_print_propcalc(x->data.binop.r);
			break;
		case Or:
			_print_propcalc(x->data.binop.l);
			printf("|");
			_print_propcalc(x->data.binop.r);
			break;
		case Imp:
			_print_propcalc(x->data.binop.l);
			printf("->");
			_print_propcalc(x->data.binop.r);
			break;
	}
}

void print_propcalc(PropCalc *x) {
	_print_propcalc(x);
	printf("\n");
}

void print_proof(Proof *x) {
	if (x == NULL) {
		printf("No proof.\n");
		return;
	}

	printf("|- ");
	print_propcalc(x->x);
}

// Constructors (terms)
PropCalc *tm_var(char a) {
	PropCalc *var = (PropCalc *)malloc(sizeof(PropCalc));

	var->tag = Var;
	var->data.var = a;

	return var;
}

PropCalc *tm_not(PropCalc *a) {
	PropCalc *data = (PropCalc *)malloc(sizeof(PropCalc));
	UnaryOp op = { .v = a };

	data->tag = Not;
	data->data.unop = op;

	return data;
}

PropCalc *tm_and(PropCalc *a, PropCalc *b) {
	PropCalc *data = (PropCalc *)malloc(sizeof(PropCalc));
	BinaryOp op = { .l = a, .r = b };

	data->tag = And;
	data->data.binop = op;

	return data;
}

PropCalc *tm_or(PropCalc *a, PropCalc *b) {
	PropCalc *data = (PropCalc *)malloc(sizeof(PropCalc));
	BinaryOp op = { .l = a, .r = b };

	data->tag = Or;
	data->data.binop = op;

	return data;
}

PropCalc *tm_imp(PropCalc *a, PropCalc *b) {
	PropCalc *data = (PropCalc *)malloc(sizeof(PropCalc));
	BinaryOp op = { .l = a, .r = b };

	data->tag = Imp;
	data->data.binop = op;

	return data;
}

// Constructors (rules)
Proof *rule_join(Proof *a, Proof *b) {
	return mk_proof(tm_and(from_proof(a), from_proof(b)));
}

Proof *rule_sep_l(Proof *a) {
	PropCalc *p = from_proof(a);
	if (p->tag != And) {
		return NULL;
	}

	return mk_proof(p->data.binop.l);
}

Proof *rule_double_tilde_intro(Proof *a) {
	PropCalc *tm_a = from_proof(a);

	free(a);

	return mk_proof(tm_not(tm_not(tm_a)));
}

Proof *rule_double_tilde_elim(Proof *a) {
	PropCalc *tm_a = from_proof(a);

	if (tm_a->tag != Not) {
		return NULL;
	}

	PropCalc *tm_aa = tm_a->data.unop.v;

	if (tm_aa->tag != Not) {
		return NULL;
	}

	Proof *prf = mk_proof(tm_aa->data.unop.v);

	free(tm_a);
	free(tm_aa);
	free(a);

	return prf;
}


Proof *rule_fantasy(PropCalc *a, Proof *(*f)(Proof *a)) {
	Proof *prfA = mk_proof(a);

	Proof *prfA2 = f(prfA);

	free(prfA);

	if (prfA2 == NULL) {
		return NULL;
	}

	PropCalc *tm_prfA2 = (PropCalc *)malloc(sizeof(PropCalc));
	memcpy(tm_prfA2, from_proof(prfA2), sizeof(PropCalc));

	free(prfA2);

	return mk_proof(tm_imp(a, tm_prfA2));
}

void eg1() {
	PropCalc *a = tm_var('a');
	PropCalc *b = tm_var('b');
	PropCalc *a_and_b = tm_and(a, b);
	print_propcalc(a_and_b);
	free_propcalc(a_and_b);
}

void eg2() {
	PropCalc *a = tm_var('a');
	PropCalc *b = tm_var('b');
	PropCalc *a_and_b = tm_and(a, b);

	Proof *prfa_and_b_to_a = rule_fantasy(a_and_b, rule_sep_l);
	print_proof(prfa_and_b_to_a);

	free_proof(prfa_and_b_to_a);
}

void eg3() {
	PropCalc *a = tm_var('a');
	PropCalc *b = tm_var('b');
	PropCalc *a_and_b = tm_and(a, b);

	Proof *prfa_and_b_to_a = rule_fantasy(a_and_b, rule_sep_l);

	Proof *double_neg = rule_double_tilde_intro(prfa_and_b_to_a);
	print_proof(double_neg);

	free_proof(double_neg);
}

void eg4() {
	PropCalc *a = tm_var('a');
	PropCalc *b = tm_var('b');
	PropCalc *a_and_b = tm_and(a, b);

	Proof *prfa_and_b_to_a = rule_fantasy(a_and_b, rule_sep_l);

	Proof *double_neg = rule_double_tilde_intro(prfa_and_b_to_a);
	Proof *double_neg_elim = rule_double_tilde_elim(double_neg);

	print_proof(double_neg_elim);
	free_proof(double_neg_elim);
}

int main() {
	eg1();
	eg2();
	eg3();
	eg4();
	return 0;
}
